mqtt/mqtt-bridge/src/persist/waking_state/waking_map.rs

```diff
@@ -0,0 +1,67 @@
+use std::{cmp::min, collections::HashMap, collections::VecDeque, task::Waker};
+
+use mqtt3::proto::Publication;
+use tracing::error;
+
+use crate::persist::{waking_state::StreamWakeableState, Key, PersistError};
+
+/// Responsible for waking waiting streams when new elements are added.
+///
+/// Exposes a get method for retrieving a count of elements in order of insertion.
+/// When elements are retrieved they are moved to the in flight collection.
+pub struct WakingMap {
+    queue: VecDeque<(Key, Publication)>,
+    in_flight: HashMap<Key, Publication>,
+    waker: Option<Waker>,
+}
+
+impl WakingMap {
+    pub fn new() -> Self {
+        let queue: VecDeque<(Key, Publication)> = VecDeque::new();
+        let in_flight = HashMap::new();
+
+        WakingMap {
+            queue,
+            in_flight,
+            waker: None,
+        }
+    }
+}
+
+impl StreamWakeableState for WakingMap {
+    fn insert(&mut self, key: Key, value: Publication) -> Result<(), PersistError> {
+        self.queue.push_back((key, value));
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+
+    fn get(&mut self, count: usize) -> Vec<(Key, Publication)> {
+        let count = min(count, self.queue.len());
+        let mut output = vec![];
+        for _ in 0..count {
+            let removed = self.queue.pop_front();
+
+            if let Some(pair) = removed {
+                output.push((pair.0.clone(), pair.1.clone()));
+                self.in_flight.insert(pair.0, pair.1);
+            } else {
+                error!("failed retrieving message from persistence");
```

Comment:
if we got a None here, we have an error in the thought process. In this case it is safer to fail fast.

Reply:
it is safer to use combinators or split_off function?
```rust
let count = min(count, self.queue.len());
let output = self.queue;
self.queue = output.split_off(count); // split_off leaves [0..count) in the output, but we want the opposite
output
```

Reply:
Regarding the first approach, it won't work as is because the creation of ```let output``` cannot move queue out of self. Also, we will need to return Vec to fit the interface so I think we will have to convert from VecDeque to Vec which might be expensive?

I chose to go with the second suggestion.